provisioner/data.js

```diff
@@ -224,32 +327,51 @@ WorkerType.prototype.testLaunchSpecs = function () {
  * so that we can create and test launch specifications before inserting
  * them into Azure.
  */
-WorkerType.createLaunchSpec = function (region, instanceType, worker, keyPrefix, provisionerId) {
+WorkerType.createLaunchSpec = function (region, instanceType, worker, keyPrefix, provisionerId, provisionerBaseUrl) {
   // These are the keys which are only applicable to a given region.
-  assert(region);
-  assert(instanceType);
-  assert(worker);
-  assert(keyPrefix);
-  assert(provisionerId);
-
-  var hasRegion = false;
-  worker.regions.forEach(function (r) {
+  assert(region, 'must specifiy region');
+  assert(instanceType, 'must specify instanceType');
+  assert(worker, 'must provide a worker object');
+  assert(keyPrefix, 'must provide key prefix');
+  assert(provisionerId, 'must provide provisioner id');
+  assert(provisionerBaseUrl, 'must provide provisioner base url');
+
+  // Find the region objects, assert if region is not found
+  var regionOverwriteObjects = {};
+  var foundRegion = false;
+  worker.regions.forEach((r) => {
     if (r.region === region) {
-      hasRegion = true;
+      // We want to make sure that we've not already found this region
+      if (foundRegion) {
+        throw new Error('Region must be unique');
+      }
+      foundRegion = true;
+      regionOverwriteObjects.launchSpec = r.launchSpec || {};
+      regionOverwriteObjects.userData = r.userData || {};
+      regionOverwriteObjects.secrets = r.secrets || {};
+      regionOverwriteObjects.scopes = r.scopes || [];
+      // Remember that we need to have an ImageId
+      assert(r.launchSpec && r.launchSpec.ImageId, 'ImageId is required in region config');
     }
   });
-  if (!hasRegion) {
-    throw new Error('workerType not configured for ' + region);
-  }
-  var hasType = false;
-  worker.instanceTypes.forEach(function (r) {
-    if (r.instanceType === instanceType) {
-      hasType = true;
+  assert(foundRegion, 'Region for workertype not found');
+
+  // Find the instanceType overwrites object, assert if type is not found
+  var instanceTypeOverwriteObjects = {};
+  var foundInstanceType = false;
+  worker.instanceTypes.forEach((t) => {
+    if (t.instanceType === instanceType) {
+      if (foundInstanceType) {
+        throw new Error('InstanceType must be unique');
```

Comment:
oh, I missed that this was for testing... then you might want to check that you don't have duplicates in `workerTypes`.
For this you can use `_.filter` instead of `_.find` which also takes an object as filter condition...
So you get:
```js
let instanceTypeInfos = _.filter(worker.instanceTypes, {instanceType});
assert(instanceTypeInfos.length === 1, 'Expected exactly one instanceType definition');
instanceTypeOverwriteObjects = _.defaults({}, instanceTypeInfos[0], {
....
```